[IOSP-254] Fix CRP Bot after board requirement changes

AliSoftware opened the pull request:
Ticket: https://babylonpartners.atlassian.net/browse/IOSP-254

### Why?

The "Release Type" and "Target Date" fields have become mandatory on CRP tickets now.

### How?

* Added the `releaseType` and `targetDate` fields to the `CRPIssueFields`
* Added a PropertyWrapper to format the date property – which allowed me to avoid having to implement `Codable`/`Content` manually property by property for types containing a `Date` property, and use a single annotation instead 🎉
* Bumped Vapor to 3.3.1 and Docker image to Swift 5.1 to be able to use PWs (no source code change)

### Tested?

* Tested the API against the FCTP JIRA board (test board with currently the same setup as the real CRP board)
* Once everything worked on FCTP board, tested on the real CRP board – and cancelled the ticket right away

### PR checklist

* [x] I've assigned this PR to myself

With :heart:

Commit:
Update gitignore for SPM/Xcode 11

Commit:
Add targetDate and releaseType fields to CRP

karabatov commented on Sources/App/CRP/JiraService+CRP.swift:
```diff
@@ -18,11 +19,20 @@ extension JiraService {
+    private static func guessTargetDate() -> Date {
+        let now = Date()
+        // Estimate time between when the CRP ticket is created and the app is released to the AppStore
+        let estimateOffset = DateComponents(day: 5)
```
This is very optimistic :D

AliSoftware replied:
Actually I put 5 here because Mark agreed that 1 week was a good enough estimate and I put 5 thinking "5 working days in a week"… but I should have put 7 instead, _especially_ since we'll soon cut the release branch on Friday morning and likely create the CRP on Friday, so at least 7 should be closer to the reality 😉 

```suggestion
        let estimateOffset = DateComponents(day: 7)
```

karabatov commented on Sources/Stevenson/CodableDate.swift:
```diff
@@ -0,0 +1,39 @@
+import Foundation
+
+@propertyWrapper
```
👍

Commit:
Bump to Swift 5.1

@propertyWrapper were introduced in Swift 5.1

karabatov reviewed (commented):
I can't really approve this because I'm not following all the changes…

AliSoftware commented:
@ilyapuchka any idea why CI continues to try to spin vapor/5.0.0 image while I updated the `.circleci/config.yml` to use 5.1?

Commit:
Bump Vapor 3.3.0 -> 3.3.1

As 3.3.1 contains fix for `import FoundationNetworking` in Swift 5.1

Commit:
Refactor ReleaseType(version:) to avoid Scanner

Since the Scanner class API changed depending on OS and Swift version, making it cumbersome otherwise to have code that can work both on Mojave/5.0 and Linux/5.1

Commit:
Add some more test cases for ReleaseType(version:)

Commit:
Update estimate offset for TargetDate to 7 days

Commit:
Code cleanup

Commit:
More generic propertyWrapper

Commit:
Merge branch 'master' into olivier/IOSP-254

AliSoftware commented:
I temporarily pointed `/crp` to the Stevenson instance corresponding to deployment of this specific PR (`babylon-stevenson-pr-38`) to let Mark give it a try and [CRP ticket was successfully created](https://babylonpartners.atlassian.net/browse/CRP-6278). 

Reverted the `/crp` slash command to point to the main instance now, but at least we can confirm that this PR fixed this issue that made it impossible to create the CRP tickets.

_Note: There are still warnings during CRP creation (some failures when setting Fix Versions on some tickets) but those are a different beast and tracked by IOSP-171 separately_

inamiy commented on Sources/Stevenson/Codable Utils/CustomCodable.swift:
```diff
@@ -0,0 +1,30 @@
+    public init(wrappedValue: Transformer.DecodedType) {
+        self.wrappedValue = wrappedValue
+    }
+}
```
This `@protocolWrapper` looks very nice!

My two cents for slight improvement is in https://gist.github.com/inamiy/eb7112bb943d8d2ed4dc3fffce4055a0 .
We can remove `TypeForCoding: Codable` constraint to make a protocol more generalised, and instead constrain on `extension` so that generalised protocol can be used in many other scenarios.

But it's OK to keep this code as is 👍

AliSoftware replied:
That's a nice idea to make the protocol be usable for other transformations besides `Codable`… but tbh I'm not a fan of the naming in your gist 😅 #NamingIsHard `convert(to: To) -> From` reads a bit strange for example given that it reads "convert to X" but in fact converts from a `To` to a `From`, feels a bit backwards on reading 🤔 

But if we find better names for that (not an easy task 😜) it's a good idea to decouple that in the future 👍

inamiy replied:
Some more ideas: `func convert / invert`, `forward / backward` , `fromTo / toFrom` ...

ilyapuchka replied:
```swift
@CustomCodable(.YMD) let date: Date

struct CustomCodable<A, B> {
    let iso: PartialIso<A, B>
    init(_ iso: PartialIso<A, B>) { ... }
}

extension PartialISO where A == String, B == Date {
  static var YMD: PartialISO {
     /* date formatter here */
  }
}
```
😉 
https://github.com/pointfreeco/swift-web/blob/f3946a64ea5a529e7023f4bf0f4e9c82d991bc82/Sources/ApplicativeRouter/PartialIso.swift

AliSoftware replied:
@ilyapuchka it's gonna work for `Codable` but not `Decodable`.
Because when decoding, we need to `init` the enclosing type passing both the `wrappedValue` (which we'd fetch from the JSON)… and the `iso` property – which we can't, since we're in the middle of `init` so we need that `.YMD` property to be `static` to be able to use it, the value of `iso` won't be provided by the JSON

That's why it's doable for `Codable` (which already has an instance that you encode, on which you have the `wrappedValue` and `iso` instance properties populated), but not for `Decodable` (for which you're in the process of creating the instance, so you don't have a value `iso`). Which led me to that solution of using `static` property and generics.

inamiy replied:
Yeah I also got stuck to use `self.iso` inside `self.init(from decoder: Decoder)` because of error: `Constant 'self.iso' used before being initialized` .

BTW alongside pointfree's impl, it seems `apply` / `unapply` is quite common naming to be used.

- https://github.com/schernichkin/partial-isomorphisms/blob/master/src/Control/Isomorphism/Partial/Prim.hs
- https://github.com/paulyoung/purescript-partial-isomorphisms/blob/v1.0.0/src/Control/Isomorphism/Partial.purs

But it might not work well if we change `Optional` return type to more accurate `Result` or `Either` to carry Error type as well.

AliSoftware replied:
Oh I like `apply/unapply`. Good use it for when I share that PW trick in my blog post 😛

Commit:
A bit of renaming on the PropertyWrapper
